### PR TITLE
feat: integrate tas to support a/b test

### PR DESCRIPTION
**Pull request by nliu-ms**

### CODE REVIEW

File: packages/vscode-extension/src/exp/index.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    public setSharedProperty(name: string, value: string): void {
+        this.sharedProperties[name] = value;
+    }
+
+    public postEvent(eventName: string, props: Map<string, string>): void {
+        let properties = { ...this.sharedProperties };
+        props.forEach((value, key) => {
+            properties[key] = value;
+        });
+        ExtTelemetry.sendTelemetryEvent(eventName, properties);
```
Comment (chagong) [inline packages/vscode-extension/src/exp/index.ts:25]: does this mean Exp has a series of its own defined telemetries?

Comment (nliu-ms) [reply packages/vscode-extension/src/exp/index.ts:25]: Yes, here is an example:
image

**Review by chagong — APPROVED** [review]

File: packages/vscode-extension/src/exp/index.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    }
+}
+
+let expService: IExperimentationService;
+
+export function getExpService(): IExperimentationService {
+    return expService
```
Comment (HuihuiWu-Microsoft) [inline packages/vscode-extension/src/exp/index.ts:32]: Semicolon

File: packages/vscode-extension/src/exp/index.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    constructor() {
+        this.sharedProperties = {};
+    }
+
+    public setSharedProperty(name: string, value: string): void {
+        this.sharedProperties[name] = value;
+    }
+
+    public postEvent(eventName: string, props: Map<string, string>): void {
```
Comment (HuihuiWu-Microsoft) [inline packages/vscode-extension/src/exp/index.ts:20]: Will Exp set shared properties and postEvent automatically?

File: packages/vscode-extension/src/exp/index.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+export function getExpService(): IExperimentationService {
+    return expService
+}
+
+export async function initialize(context: vscode.ExtensionContext) {
+    expService = await getExperimentationServiceAsync(
+        pkg.name,
+        pkg.version,
+        TargetPopulation.Public,
```
Comment (HuihuiWu-Microsoft) [inline packages/vscode-extension/src/exp/index.ts:39]: Does this mean public users will be in the test?

File: packages/vscode-extension/src/exp/index.ts (hunk 3 above)
Comment (nliu-ms) [reply packages/vscode-extension/src/exp/index.ts:20]: Yes, these two functions will be called automatically, as this telemetry object is passed to ExperimentationService.

File: packages/vscode-extension/src/exp/index.ts (hunk 4 above)
Comment (nliu-ms) [reply packages/vscode-extension/src/exp/index.ts:39]: I think our target is testing changes for public population.

**Review by HuihuiWu-Microsoft — APPROVED** [review]

**Review by HuihuiWu-Microsoft — APPROVED** [review]

File: packages/vscode-extension/src/exp/index.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            pkg.version,
+            TargetPopulation.Public,
+            new ExperimentationTelemetry(),
+            context.globalState
+        );
+    }
+
+    class ExperimentationTelemetry implements IExperimentationTelemetry {
+        public setSharedProperty(name: string, value: string): void {
+            sharedProperties[name] = value;
```
Comment (daytonellwanger) [inline packages/vscode-extension/src/exp/index.ts:33]: These shared properties need to be attached to all telemetry events your extension sends - it's how we can tie an event to a specific experiment group and create a scorecard. It looks like they're just being added to `sharedProperties` and never used?

Comment (daytonellwanger) [reply packages/vscode-extension/src/exp/index.ts:33]: Oh I see - they're being attached to all events sent via `postEvent`. They need to be attached to those events *and* all other events your extension sends.

Comment (nliu-ms) [reply packages/vscode-extension/src/exp/index.ts]: Got it, thanks!

File: packages/vscode-extension/src/exp/index.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        public setSharedProperty(name: string, value: string): void {
+            ExtTelemetry.addSharedProperty(name, value);
+        }
+    
+        public postEvent(eventName: string, props: Map<string, string>): void {
+            let properties: {[key: string]: string} = {};
+            props.forEach((value, key) => {
+                properties[key] = value;
+            });
+            ExtTelemetry.sendTelemetryEvent(eventName, properties);
```
Comment (HuihuiWu-Microsoft) [inline packages/vscode-extension/src/exp/index.ts:36]: One question, the props here are not the shared properties that you attached  to vscode telemetry reporter?

**Review by swatDong — APPROVED** [review]

File: packages/vscode-extension/src/exp/index.ts (hunk 6 above)
Comment (nliu-ms) [reply packages/vscode-extension/src/exp/index.ts:36]: here is the example of the props, other shared properties will be added by the reporter.
image

**Review by HuihuiWu-Microsoft — APPROVED** [review]

File: packages/vscode-extension/src/exp/index.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import * as vscode from "vscode";
+import { IExperimentationService, IExperimentationTelemetry, TargetPopulation, getExperimentationServiceAsync } from "vscode-tas-client";
+import * as pkg from "./../../package.json";
+import { ExtTelemetry } from "./../telemetry/extTelemetry"
+
+export namespace exp {
```
Comment (1openwindow) [inline packages/vscode-extension/src/exp/index.ts:9]: what is this new folder for? what is it stand for?

**Pull request merged**